homeassistant/components/hunterdouglas_powerview/diagnostics.py

```diff
@@ -0,0 +1,114 @@
+"""Diagnostics support for Powerview Hunter Douglas."""
+from __future__ import annotations
+
+from typing import Any
+
+import attr
+
+from homeassistant.components.diagnostics import async_redact_data
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+from homeassistant.helpers.device_registry import DeviceEntry
+
+from .const import (
+    COORDINATOR,
+    DEVICE_INFO,
+    DEVICE_MAC_ADDRESS,
+    DEVICE_SERIAL_NUMBER,
+    DOMAIN,
+    PV_HUB_ADDRESS,
+)
+from .coordinator import PowerviewShadeUpdateCoordinator
+
+REDACT_CONFIG = {
+    CONF_HOST,
+    DEVICE_MAC_ADDRESS,
+    DEVICE_SERIAL_NUMBER,
+    PV_HUB_ADDRESS,
+    "configuration_url",
+}
+
+
+async def async_get_config_entry_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a config entry."""
+    return _async_get_diagnostics(hass, entry)
+
+
+async def async_get_device_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry, device: DeviceEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a device entry."""
+    return _async_get_diagnostics(hass, entry, device)
+
+
+@callback
+def _async_get_diagnostics(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    device: DeviceEntry | None = None,
+) -> dict[str, Any]:
+    """Return diagnostics for a config entry."""
+
+    pv_data = hass.data[DOMAIN][entry.entry_id]
+    coordinator: PowerviewShadeUpdateCoordinator = pv_data[COORDINATOR]
+    shade_data = coordinator.data.get_all_raw_data()
+    hub_info = async_redact_data(pv_data[DEVICE_INFO], REDACT_CONFIG)
+
+    data = {"hub_info": hub_info, "shade_data": shade_data}
+
+    if device:
+        data["device_info"] = _async_device_as_dict(hass, device)
+        # try to match on name to restrict to shade if we can
+        # otherwise just return all shade data
+        # shade name is unique in powerview
+        for shade in shade_data:
+            if shade_data[shade]["name_unicode"] == device.name:
+                data["shade_data"] = shade_data[shade]
+
+    else:
+        device_registry = dr.async_get(hass)
+        data.update(
+            device_info=[
+                _async_device_as_dict(hass, device)
+                for device in dr.async_entries_for_config_entry(
+                    device_registry, entry.entry_id
+                )
+            ],
+        )
+
+    return data
```

Comment:
```suggestion
async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    data = _async_get_diagnostics(hass, entry)
    device_registry = dr.async_get(hass)
    data.update(
        device_info=[
            _async_device_as_dict(hass, device)
            for device in dr.async_entries_for_config_entry(
                device_registry, entry.entry_id
            )
        ],
    )
    return data


async def async_get_device_diagnostics(
    hass: HomeAssistant, entry: ConfigEntry, device: DeviceEntry
) -> dict[str, Any]:
    """Return diagnostics for a device entry."""
    data = _async_get_diagnostics(hass, entry)
    data["device_info"] = _async_device_as_dict(hass, device)
    # try to match on name to restrict to shade if we can
    # otherwise just return all shade data
    # shade name is unique in powerview
    shade_data = data["shade_data"]
    for shade in shade_data:
        if shade_data[shade]["name_unicode"] == device.name:
            data["shade_data"] = shade_data[shade]
    return data


@callback
def _async_get_diagnostics(
    hass: HomeAssistant,
    entry: ConfigEntry,
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    pv_data = hass.data[DOMAIN][entry.entry_id]
    coordinator: PowerviewShadeUpdateCoordinator = pv_data[COORDINATOR]
    shade_data = coordinator.data.get_all_raw_data()
    hub_info = async_redact_data(pv_data[DEVICE_INFO], REDACT_CONFIG)
    return {"hub_info": hub_info, "shade_data": shade_data}
```